### ISSUE TITLE
[feat] platform/arch-specific resources

### ISSUE THREAD

**Issue by zhangsan946** [tauri-apps/tauri#8501]
### Describe the problem

In additional to executable files, static/dynamic library files are also platform & arch dependent. Need a way to embed external binaries for other extentions such as: .dll, .so,  beside executable files.

### Describe the solution you'd like

Support externalBin setup as following:
```
"externalBin": [
    "binaries/my-sidecar"
]
```
And in additional:
```
"externalBin": [
    {path: "binaries/my-sidecar-dll", ext: "dll"}
]
```

**Comment by FabianLars** [comment]
I'd say the resources feature is what you're asking for <https://tauri.app/v1/guides/building/resources/>. They can be platform specific via https://tauri.app/v1/api/config#platform-specific-configuration.

Your pr also only works on windows because linux and macos typically don't have library files next to executables.

**Comment by FabianLars** [comment]
whoops, didn't mean to close this before you have a chance to respond

**Comment by zhangsan946** [comment]
Hi @FabianLars, thanks for the reponse. The resource feature won't help bescause the library file is also arch dependent like the excutable file. Please take the wintun.dll file as an example https://www.wintun.net/builds/wintun-0.14.1.zip. It has to keep the name of `wintun.dll` and it has different version for x86, amd64, arm.. unless resource feature could be not only platform specific  but also arch specific.
On linux, library search path is `LD_LIBRARY_PATH` specified path,  executable's rpath, /lib folder. User could set the LD_LIBRARY_PATH if the executable's rpath is not set to be the same folder to executables.

**Comment by FabianLars** [comment]
> The resource feature won't help bescause the library file is also arch dependent like the excutable file

right, maybe we could expand the platform specific configs to also be arch specific (optionally) but i can't think of a different usecase than this issue so maybe that's a bit too much. Using the `--config`/`-c` flag works but having to use a manual step for this feels awkward too.

> On linux, library search path is LD_LIBRARY_PATH specified path, executable's rpath, /lib folder. User could set the LD_LIBRARY_PATH if the executable's rpath is not set to be the same folder to executables.

That's why i said your solution doesn't work. In the .deb bundle the sidecars will be installed into /usr/bin where .so files don't belong. In the .appimage bundle the rpath and iirc the ld_library_path are set but either way appimages also have an internal `/usr/bin/` folder where it wouldn't be as bad to have .so files in but still unusual.

----

So yes, i agree that we need _something_ but i also think it needs a bit more thinking. Abusing the externalBin/sidecar config for this doesn't seem right.

**Comment by minht11** [comment]
I have use case for this.
I am building Windows only application on a Mac (ironic I know). UI does not need native lib, so I could build myself abstraction which calls sidecar only when on Windows and still works for most other stuff on other platforms. Right now that is not possible.

Allowing platforms to be optional would be great.

**Comment by FabianLars** [comment]
@minht11 Your use-case sounds like it's already possible: https://tauri.app/v1/api/config#platform-specific-configuration

This issue here is about specific configs for platform+architecture combinations, so for example different configs for `windows-x86_64` and `windows-aarch64`. If you only build the app for windows x64 (so like 99% of windows users) the current system should work fine.

**Comment by brand-it** [comment]
I ran into something like this with dependent packages for Mac. The ref PR(https://github.com/tauri-apps/tauri/pull/8502) was kinda of what I was thinking for a solution to work around needing `dylib` for Mac. Thankfully Mac binary files normal don't have a .exe or .sh added to the end so I got round the problem by simply not just adding the `dylib` to the `externalBin`. Worked great, Down side is I had a need to build that same binary for windows and they needed you guessed it `dll` files. The down side is the code alway adds the `.exe` extension to the files which brakes the binary dependency.

`binaries/driverio.dll-x86_64-pc-windows-msvc` 
`binaries/driverio.dylib-aarch64-apple-darwin`

I have been trying to think of a way to use [tauri](https://github.com/tauri-apps/tauri) to make this work but keep running into the issue with it adding the `.exe` to the file name

https://github.com/tauri-apps/tauri/blob/4633705da7a5ea966ca6bfe7c8c152ab18a4b5ce/crates/tauri-utils/src/resources.rs#L51-L52

I even go for a solution that lets me include dependencies or something but I have a felling the best solution right now is to build a custom script that renames the files as a post process for building windows. 

Correct me if I'm got this wrong but that is what I'm thinking this thread is about? 


PS: 
_Also as a side note great work on this tool, making windows and Mac work together is not easy and I hate to have to think of all these edge cases_

**Comment by FabianLars** [comment]
> Correct me if I'm got this wrong but that is what I'm thinking this thread is about?

Like mentioned a few times here, externalBin is not meant for libraries, so yes this is the right place for what you need.

**Comment by haexhub** [comment]
sorry to bother you again, but could you please briefly summarize how to best integrate platform and arch specific libraries into a tauri project?
Something like externalBinaries just for libraries would be nice I guess

**Comment by nandor23** [comment]
Unfortunately, there's no clean built-in solution with sidecars yet, but here are two potential workarounds:

<details>
<summary>Approach 1: Sidecar + Resources + PATH manipulation</summary>

### Config:
```json
// tauri.conf.json
{
  "bundle": {
    "externalBin": ["binaries/my-tool"]
  }
}
```
```json
// tauri.windows.conf.json
{
  "bundle": {
    "resources": [
      "binaries/my-lib.dll",
      "binaries/another-lib.dll"
    ]
  }
}
```

### Rust:
```rust
let resource_path = app.path().resource_dir()?;
let binaries_path = resource_path.join("binaries");
let mut command = app.shell().sidecar("my-tool")?;
command = command.env("PATH", format!("{};{}", 
    binaries_path.to_string_lossy(), 
    std::env::var("PATH").unwrap_or_default()
));
```

### Issues:
Requires manual PATH manipulation

</details>

<details>
<summary>Approach 2: All resources + direct command path</summary>

### Config:
```json
// tauri.windows.conf.json
{
  "bundle": {
    "resources": [
      "binaries/windows/my-tool.exe",
      "binaries/windows/my-lib.dll"
    ]
  }
}
```
```json
// capabilities/default.json
{
  "$schema": "../gen/schemas/desktop-schema.json",
  "identifier": "default",
  "description": "Capability for the main window",
  "windows": ["main"],
  "permissions": [
    "core:default",
    "shell:default",
    {
      "identifier": "shell:allow-execute",
      "allow": [
        {
          "name": "my-tool",
          "cmd": "binaries/windows/my-tool.exe",
          "args": true
        }
      ]
    }
  ]
}
```

### Rust:
```rust
let resource_path = app.path().resource_dir()?;
let tool_path = resource_path.join("binaries/windows/my-tool.exe");
app.shell().command(tool_path.to_string_lossy().to_string())
    .args(args)
    .output()
    .await?;
```

### Issues:
- Requires manual path construction for each OS (separate resource paths needed for windows/macos/linux)
- Requires custom shell permission configuration for the command

</details>

---

Approach 1 relies on PATH manipulation to help the sidecar find its libraries. Approach 2 places the executable and libraries in the same directory so the OS can find them naturally. Approach 1 would be ideal if Tauri automatically resolved and bundled sidecar dependencies without manual PATH configuration.

I am not happy with the current DX, but the 2nd approach is more maintainable. 

Some `.exe` files need `.dll`s to execute, like Android Debug Bridge on Windows, which is dependent on 2 other DLL files. Without the workarounds, the executable fails to run because it cannot locate its dependencies.
